File: src/main/java/com/canefe/story/NPCDataManager.java
package com.canefe.story;

import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class NPCDataManager {

    private static NPCDataManager instance;

    private final JavaPlugin plugin;
    private final Map<String, NPCData> npcDataMap;
    private final File npcDirectory;

    private NPCDataManager(JavaPlugin plugin) {
        this.plugin = plugin;
        this.npcDirectory = new File(plugin.getDataFolder(), "npcs");
        this.npcDataMap = new HashMap<>();

        if (!npcDirectory.exists()) {
            npcDirectory.mkdirs(); // Create the directory if it doesn't exist
        }
    }

    public static NPCDataManager getInstance(Story plugin) {
        if (instance == null) {
            instance = new NPCDataManager(plugin);
        }
        return instance;
    }


    public File getNPCDirectory() {
        return npcDirectory;
    }


    public FileConfiguration loadNPCData(String npcName) {
        File npcFile = new File(getNPCDirectory(), npcName + ".yml");
        if (!npcFile.exists()) {
            return new YamlConfiguration(); // Return an empty configuration if no file exists
        }
        FileConfiguration config = YamlConfiguration.loadConfiguration(npcFile);

        return YamlConfiguration.loadConfiguration(npcFile);
    }


    public void saveNPCFile(String npcName, FileConfiguration config) {
        File npcFile = new File(npcDirectory, npcName + ".yml");
        try {
            config.save(npcFile);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void deleteNPCFile(String npcName) {
        File npcFile = new File(npcDirectory, npcName + ".yml");
        if (npcFile.exists()) {
            npcFile.delete();
        }
    }
}

File: src/main/kotlin/com/canefe/story/npc/data/NPCDataManager.kt
package com.canefe.story.npc.data

import com.canefe.story.Story
import com.canefe.story.conversation.ConversationMessage
import net.citizensnpcs.api.CitizensAPI
import net.citizensnpcs.api.npc.NPC
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException

class NPCDataManager private constructor(private val plugin: JavaPlugin) {
    private val npcDataMap: MutableMap<String, NPCData> = HashMap()
    val npcDirectory: File = File(plugin.dataFolder, "npcs").apply {
        if (!exists()) {
            mkdirs() // Create the directory if it doesn't exist
        }
    }

    /**
     * Gets a list of all NPC names by scanning the NPC directory for YAML files.
     *
     * @return A list of all NPC names without the .yml extension
     */
    fun getAllNPCNames(): List<String> {
        val npcNames = ArrayList<String>()

        val files = npcDirectory.listFiles { _, name -> name.endsWith(".yml") }
        files?.forEach { file ->
            val fileName = file.name
            // Remove the .yml extension to get the NPC name
            val npcName = fileName.substring(0, fileName.length - 4)
            npcNames.add(npcName)
        }

        return npcNames
    }

    fun loadNPCData(npcName: String): FileConfiguration {
        val npcFile = File(npcDirectory, "$npcName.yml")
        if (!npcFile.exists()) {
            return YamlConfiguration() // Return an empty configuration if no file exists
        }

        return YamlConfiguration.loadConfiguration(npcFile)
    }

    fun saveNPCFile(npcName: String, config: FileConfiguration) {
        val npcFile = File(npcDirectory, "$npcName.yml")
        try {
            config.save(npcFile)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun getNPC(npcName: String): NPC? {
        val npcFile = File(npcDirectory, "$npcName.yml")
        if (!npcFile.exists()) {
            return null // Return null if no file exists
        }

        // Try to find npc by name in citizens registry
        return CitizensAPI.getNPCRegistry().find { it.name.equals(npcName, ignoreCase = true) }
    }

    fun deleteNPCFile(npcName: String) {
        val npcFile = File(npcDirectory, "$npcName.yml")
        if (npcFile.exists()) {
            npcFile.delete()
        }
    }

    companion object {
        private var instance: NPCDataManager? = null

        @JvmStatic
        fun getInstance(plugin: Story): NPCDataManager {
            return instance ?: NPCDataManager(plugin).also { instance = it }
        }
    }
}